ufl/core/external_operator.py

```diff
@@ -0,0 +1,278 @@
+                s = ref_coefficient.ufl_shape
+                for i, e in enumerate(self.derivatives):
+                    s += self.ufl_operands[i].ufl_shape * e
+                original_function_space = self._make_function_space(s, sub_element=ref_coefficient.ufl_element())
+                ufl_shape = original_function_space.ufl_element().reference_value_shape()
+            else:
+                ufl_shape = ref_coefficient.ufl_element().reference_value_shape()
+                original_function_space = ref_coefficient.ufl_function_space()
+        else:
+            ufl_shape = ref_coefficient.ufl_element().reference_value_shape()
+            original_function_space = ref_coefficient.ufl_function_space()
+            self.derivatives = (0,) * len(self.ufl_operands)
+
+        if coefficient is None:
+            coefficient = Coefficient(original_function_space, count=count)
+        elif not isinstance(coefficient, Coefficient):
+            raise TypeError('Expecting a Coefficient and not %s', type(coefficient))
+        self._coefficient = coefficient
+        self.ufl_shape = ufl_shape
+        self._original_function_space = self.coefficient.ufl_function_space()
+
+        self._extop_dependencies = [self, ]
+
+        if self.derivatives == (0,) * len(self.ufl_operands):
+            self._extop_master = self
+
+    @property
+    def coefficient(self):
+        return self._coefficient
+
+    def _extract_coeffs_and_args(self, operands):
+        from ufl.algorithms.analysis import extract_arguments
+        args = tuple((e, is_adjoint) for e, is_adjoint in operands if len(extract_arguments(e)) != 0)
+        ops = tuple((e, is_adjoint) for e, is_adjoint in operands if (e, is_adjoint) not in args)
+        return ops, args
+
+    def arguments(self):
+        return self._arguments
+
+    def action_args(self):
+        return self._action_args
+
+    @property
+    def _extop_items(self):
+        return self.ufl_operands + (self.coefficient,) + tuple(arg for arg, _ in self._arguments)
+
+    def count(self):
+        return self._count
+
+    @property
+    def _count(self):
+        return self.coefficient._count
+
+    @_count.setter
+    def _count(self, value):
+        self.coefficient._count = value
+
+    @property
+    def ufl_shape(self):
+        "Return the UFL shape of self.coefficient."
+        return self.coefficient._ufl_shape
+
+    @ufl_shape.setter
+    def ufl_shape(self, shape):
+        self.coefficient._ufl_shape = shape
+
+    def ufl_function_space(self):
+        return self._ufl_function_space
+
+    @property
+    def _original_ufl_function_space(self):
+        "Get the function space of this coefficient."
+        return self.coefficient._ufl_function_space
+
+    @_original_ufl_function_space.setter
+    def _original_ufl_function_space(self, func_space):
+        "Get the function space of this coefficient."
+        self.coefficient._ufl_function_space = func_space
```

Comment:
In place modification of things is bad. Especially because now `ufl_function_space` and `original_ufl_function_space` can end up going out of sync. Why do you need this?

Reply:
We don’t need this, that should have been removed! I removed it.